test/python/provider-upgrade-command/test.ts

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) HashiCorp, Inc
+// SPDX-License-Identifier: MPL-2.0
+import { TestDriver } from "../../test-helper";
+
+describe("provider upgrade command", () => {
+  let driver: TestDriver;
+
+  describe("pre-built", () => {
+    describe("pipenv", () => {
+      beforeAll(async () => {
+        driver = new TestDriver(__dirname, {
+          CDKTF_DIST: "",
+          DISABLE_VERSION_CHECK: "true",
+        }); // reset CDKTF_DIST set by run-against-dist script & disable version check as we have to use an older version of cdktf-cli
+        await driver.setupPythonProject({
+          init: { additionalOptions: "--cdktf-version 0.10.4" },
+        });
+      });
+
+      test("updates pre-built provider using pipenv", async () => {
+        await driver.exec("pipenv", [
+          "install",
+          `cdktf-cdktf-provider-random==0.2.55`,
+        ]);
+        expect(driver.readLocalFile("Pipfile")).toContain(
+          `cdktf-cdktf-provider-random = "==0.2.55"`
+        );
+
+        await driver.exec("cdktf", ["provider", "upgrade", "random@=3.2.0"]);
+        expect(driver.readLocalFile("Pipfile")).toContain(
+          `cdktf-cdktf-provider-random = "~=0.2.64"`
+        );
+      });
+    });
+
+    describe("pip", () => {
+      beforeAll(async () => {
+        driver = new TestDriver(__dirname, {
+          CDKTF_DIST: "",
+          DISABLE_VERSION_CHECK: "true",
+        }); // reset CDKTF_DIST set by run-against-dist script & disable version check as we have to use an older version of cdktf-cli
+        await driver.setupPythonProject({
+          init: { additionalOptions: "--cdktf-version 0.10.4" },
+        });
+
+        driver.copyFile("cdktf-pip.json", "cdktf.json");
```

Comment:
Not feedback: Ah, we don't have a way of pointing to a custom cdktf.json file? Would be useful in the future at least for testing etc. perhaps.

Reply:
Yeah we heard this sporadically from customers as well. Having multiple makes it a bit harder to think about what a cdktf project is. With the current implementation it's "a directory containing a cdktf.json" and I think that simplicity is okay for now.